torch_xla_py/xla_dist.py

```diff
@@ -402,3 +409,364 @@ def get_cluster(self):
+      script.append(['.', '/etc/profile'])
+      if self.docker_image:
+        script.append(self._docker_run_cmd(cmd))
+      else:
+        if self.conda_env:
+          script.append(['conda', 'activate', self.conda_env])
+        script.append(cmd)
+
+      # ex. script_body = 'conda activate pytorch; python train.py'
+      script_body = '; '.join([' '.join(command) for command in script])
+      os.makedirs(os.path.dirname(script_path), exist_ok=True)
+      with open(script_path, 'w') as f:
+        f.write(script_body)
+      subprocess.call(['chmod', '+x', script_path])
+      worker_script_map[self._cluster._client_workers[i]] = script_path
+
+    return worker_script_map
+
+  def _scp_scripts(self, script_map):
+
+    def _gcloud_scp(script_path, client_worker):
+      scp_cmd = self._build_scp_cmd(script_path, os.path.basename(script_path),
+                                    client_worker)
+      proc = subprocess.Popen(
+          scp_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+      self._stream_logs(proc, client_worker)
+
+    threads = []
+    for i, client_worker in enumerate(script_map):
+      if i == 0:
+        # ssh keygen single time
+        _gcloud_scp(script_map[client_worker], client_worker)
+        continue
+      thread = threading.Thread(
+          target=_gcloud_scp, args=(
+              script_map[client_worker],
+              client_worker,
+          ))
+      thread.daemon = True
+      thread.start()
+      threads.append(thread)
+
+    for thread in threads:
+      thread.join()
+
+  def _cleanup(self, script_map):
+
+    def _cleanup_worker(script_path, client_worker):
+      rm_script = ['rm', '~/{}'.format(os.path.basename(script_path))]
+      self._build_and_run_ssh(rm_script, client_worker)
+      subprocess.call(['rm', script_path])
+      if self.docker_image:
+        rm_container = ['docker', 'rm', '-f', self.docker_container]
+        self._build_and_run_ssh(rm_container, client_worker)
+      rm_proc = ['pkill', '-u', self.DEFAULT_USER_NAME]
+      self._build_and_run_ssh(rm_proc, client_worker, log=False)
+
+    threads = []
+    for client_worker in script_map:
+      thread = threading.Thread(
+          target=_cleanup_worker,
+          args=(
+              script_map[client_worker],
+              client_worker,
+          ))
+      thread.start()
+      threads.append(thread)
+    for thread in threads:
+      thread.join()
+
+  def _start_run(self, script_map):
+
+    def _run_script(script_path, client_worker):
+      self._build_and_run_ssh(
+          ['~/{}'.format(os.path.basename(script_path))], client_worker)
+
+    threads = []
+    for client_worker in script_map:
+      thread = threading.Thread(
+          target=_run_script, args=(
+              script_map[client_worker],
+              client_worker,
+          ))
+      thread.daemon = True
+      thread.start()
+      threads.append(thread)
+
+    try:
+      for thread in threads:
```

Comment:
The whole multiple thread wait is in many places.
As it is, a Ctrl^C here won't stop running processes (the Popen ones) to stop, and unwind correctly.
We will likely need an event passed down to the log streamers, which, within the core readline+log loop, check whether the event is set.
If set, quit the loop, kill the process, and exit the thread.

Then on the Ctrl^C path, we set the event, and then go back waiting for the threads to notice it and unwind.

I am OK this to be done in an (immediately) follow up PR, since we have been iterating this for a while now ;)

Reply:
That sounds good to me. On the cleanup path we also have an explicit `pkill` for the user under which all the remote cmds are issued so I believe the `Popen` processes to get killed (in addition to any other remote processes under same user that the main remote process may have started). But yes, I agree I should pass down the event to the log streamers, and if set stop the streaming, and after the streaming we have `proc.kill()`. I'll do this in a followup as suggested. Thanks!

Reply:
Not questioning the `pkill` ... but if you do not notice you have to quit, you will never issue it (or issue it much much later when the job reaches completion).

Reply:
Right, I'll plumb through the `threading.event` and set that at `KeyboardInterrupt` time to kill all the `Popen` objects.